tools/gcsbackup/GCSbackup.sh

```diff
@@ -0,0 +1,124 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Please note: GCSbackup should only run inside a GCloud VM instance.
+#
+# GCSbackup is a tool to create and maintain a tagged .tar backup archive
+# for a given local path and copy it to a GCS bucket. It can be configured to
+# maintain a maximum of n backups for specified tag. It automatically
+# deletes older backups with the same tag.
+#
+# On GCS, the .tar is copied to a qualified path that is unique to the VM
+# where this script is running, the username, path, tag, and timestamp.
+
+USAGE='USAGE:
+
+  ./GCSbackup.sh [OPTION]...
+
+OPTIONS:
+
+  -n, --num-backups   Number of backups to keep. Default is 10
+  -b, --bucket        Name of GCS bucket to store backups in. Default is "{project-id}_datalab-backups"
+                      Follow the bucket naming guidelines here: https://cloud.google.com/storage/docs/naming
+  -p, --path          Path to backup. Default is current directory
+  -t, --tag           Tag to make grouping similar backups easy. Default is "backup"
+'
+
+while [[ $# -gt 1 ]]; do
+  key="$1"
+  case $key in
+      -n|--num-backups)
+        num_backups="$2"
+        shift
+        ;;
+      -b|--bucket)
+        gcs_bucket="$2"
+        shift
+        ;;
+      -p|--path)
+        backup_path="$2"
+        shift
+        ;;
+      -t|--tag)
+        tag="$2"
+        shift
+        ;;
+      --default)
+        DEFAULT=YES
+        shift
+        ;;
+      *)
+        echo "Bad arguments found: ${key}"
+        echo "${USAGE}"
+        exit 1
+      ;;
+  esac
+  shift   # skip option value
+done
+
+timestamp=$(date "+%Y%m%d%H%M%S")
+machine_id=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/id" -H "Metadata-Flavor: Google" || echo "")
+project_id=$(gcloud info |tr -d '[]' | awk '/project:/ {print $2}')
+default_bucket="${project_id}_datalab-backups"
+tag="${tag:-backup}"
+num_backups=${num_backups:-10}
+gcs_bucket=${gcs_bucket:-$default_bucket}
+backup_path=`readlink -f "${backup_path:-.}"`
+
+echo "tag: ${tag}"
+echo "backups to keep: ${num_backups}"
+echo "backup path: ${backup_path}"
+echo "gcs bucket:  ${gcs_bucket}"
+echo
+
+if [[ -z $machine_id ]]; then
+  echo "GCSbackup can only run on a GCloud VM instance"
+  exit 1
+fi
+
+# test and create bucket if necessary
+gsutil ls gs://${gcs_bucket} &>/dev/null || {
+  echo "Bucket '${gcs_bucket}' was not found. Creating it.."
+  gsutil mb gs://"${gcs_bucket}"
+}
+
+# create an archive of the backup path
+archive_name=$(mktemp -d)"/archive.tar"
+echo "Creating archive: $archive_name"
+tar -cf ${archive_name} "${backup_path}" || {
+  echo "Failed creating the backup archive"
+  exit 1
+}
+
+backup_id="${gcs_bucket}${backup_path}/${machine_id}-${USER}-${tag}-${timestamp}"
```

Comment:
Will backup_path always begin with a /?

We should probably either check for that or add the slash here and in the all_backups line below.